### PR TITLE
feat: follow spot via inline button

### PR DESCRIPTION
**Pull request by TaToTanWeb**
Possible implementation of the feature suggested in #56. It creates a new SQL table that will later contain user IDs of the people who choose to follow the spot: future comments on that spot would then be copied in the user's private chat with the bot.
**Note:** the python-telegram-bot library in requirements.txt is 13.15 because the field `message_thread_id` in `Message` was not available in 13.14.

### CODE REVIEW

**Review by Helias — APPROVED** [review]

**Comment by Helias** [comment]
@TaToTanWeb, unbelievable job 🚀 

Check the pylint errors in the pipeline please

File: modules/handlers/follow_spot.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        # Forget the stored data
+        DbManager.delete_from(
+            table_name="user_follow",
+            where="user_id = %s and message_id = %s",
+            where_args=(info.user_id, message_id)
+        )
+        
+        info.bot.send_message(
+            chat_id = info.user_id,
+            text = "Non stai più seguendo questo spot.",
```
Comment (TendTo) [inline modules/handlers/follow_spot.py:44]: Immagino qui volessi usare `answer_text`

File: modules/handlers/follow_spot.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                chat_id = info.user_id,
+                from_chat_id = info.chat_id,
+                message_id = message_id,
+                disable_notification = True
+            )
+
+            # Add an explanation to why the message was forwarded?
+            info.bot.send_message(
+                chat_id = info.user_id,
+                text = "Stai seguendo questo spot.",
```
Comment (TendTo) [inline modules/handlers/follow_spot.py:63]: E anche qui

File: modules/utils/keyboard_util.py [hunk 3]
```diff
@@ -147,32 +147,24 @@ def get_paused_kb(page: int, items_per_page: int) -> InlineKeyboardMarkup:
...
-        if len(keyboard) > 0:
-            keyboard[-1].append(report_button)
-        else:
-            keyboard.append([report_button])
-
-    return InlineKeyboardMarkup(keyboard) if keyboard else None
+    return InlineKeyboardMarkup([
+        *([[InlineKeyboardButton(
+            "🚩 Report",
+            callback_data="meme_report_spot"
```
Comment (TendTo) [inline modules/utils/keyboard_util.py:161]: Manterrei comunque il vecchio sistema che supporta anche le reaction manuali gestite dal bot. So che non lo stiamo usando più, però è comunque una feature in più. In alternativa, avrebbe senso rimuovere del tutto la funzionalità, inclusi tutti le impostazioni che si tira dietro

File: modules/handlers/follow_comment.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        update: update event
+        context: context passed by the handler
+    """
+    info = EventInfo.from_message(update, context)
+
+    if info.chat_id == Config.meme_get('channel_group_id'):
+        # Get the spot's message_id
+        reply_to_message_id = info.message.message_thread_id
+        # Get a list of users who are following the spot
+        result = DbManager.select_from(
```
Comment (TendTo) [inline modules/handlers/follow_comment.py:23]: Io personalmente sono un fan dell'approccio ORM (Object Relational Mapping) per quanto riguarda le interazioni con il database. Però si tratta di preferenze personali

**Comment by Helias** [comment]
I strongly like this feature and I tested it on localhost.

I would like to suggest two improvements:
- add the post link when it says "stai seguendo questo spot", so the user can reach easily the post clicking on it
- do not update me if I write a comment, you have the `user_id`, you could make a match with the chat id of the comment and with the user who is following it

image

File: modules/handlers/follow_comment.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            table_name = "user_follow",
+            select = "user_id, private_message_id",
+            where = "message_id = %s",
+            where_args = (reply_to_message_id, )
+        )
+
+        # Send them an update about the new comment
+        for user in result:
+            info.message.copy(chat_id=user['user_id'], reply_to_message_id=user['private_message_id'])
```
Comment (Helias) [inline modules/handlers/follow_comment.py:32]: Do not send the notification when the comment author is the follower
```suggestion
        # Send them an update about the new comment
        for user in result:
            if not user['user_id'] == info.message.from_user.id:
                info.message.copy(chat_id=user['user_id'], reply_to_message_id=user['private_message_id'])
```

File: modules/handlers/follow_spot.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        )
+        
+        info.bot.send_message(
+            chat_id = info.user_id,
+            text = "Non stai più seguendo questo spot.",
+            reply_to_message_id = message_id,
+            disable_notification = True
+        )
+    else:
+        answer_text = "Stai seguendo questo spot."
```
Comment (Helias) [inline modules/handlers/follow_spot.py:49]: you could add something like 
```suggestion
        post_url = f"https://t.me/c/{str(info.chat_id).replace('-100', '')}/{info.message_id}"
        answer_text = f"Stai seguendo questo spot. {post_url}"
```

You could also replace that text URL into an icon ➡️  with the href="post_url"...
or put a telegram button with the arrow

Comment (Helias) [reply modules/handlers/follow_spot.py:49]: p.s. you could take the info.chat_id variable from settings.yaml `channel_group_id`, but it's the same

**Comment by TaToTanWeb** [comment]
Thank you! The code should be better now.
p.s. I can't seem to find the "trailing whitespace" from the last CodeFactor analysis 🤔

**Comment by Helias** [comment]
some tests are failing

**Comment by Helias** [comment]
the test bot class is not working anymore @tendto

**Comment by Helias** [comment]
testing it locally, it seems working, but we should fix the tests

**Comment by TendTo** [comment]
I think that the new telegram library version has some new features my mocked class does not support. As soon as I have some free time, I will try and update it.

**Comment by Helias** [comment]
@TaToTanWeb align with this repository your fork please

**Pull request merged**